decoder/stream_decoder.go

```diff
@@ -19,46 +19,61 @@ package decoder
 
 import (
 	"bytes"
+	"encoding/json"
 	"io"
-	"io/ioutil"
 )
 
 func NewNDJSONStreamReader(reader *LineReader) *NDJSONStreamReader {
-	return &NDJSONStreamReader{reader: reader}
+	var r NDJSONStreamReader
+	r.Reset(reader)
```

Comment:
The `LineReader` and the underlying `bufio.Reader` are created and managed outside the `NDJSONStreamReader` but also inside when e.g. reading. What do you think about moving the whole administration inside the `NDJSONStreamReader` e.g. by doing the following:
```

type NDJSONStreamReader struct {
	buf              *bufio.Reader
	reader           *LineReader
	isEOF            bool
	latestLine       []byte
	latestLineReader bytes.Reader
	decoder          *json.Decoder
}
func NewNDJSONStreamReader(reader io.Reader, size int) *NDJSONStreamReader {
	var r NDJSONStreamReader
	r.buf = bufio.NewReaderSize(reader, size)
	r.reader = NewLineReader(r.buf, size)
	r.isEOF = false
	r.latestLine = nil
	r.latestLineReader.Reset(nil)
	r.resetDecoder()
	return &r
}
func (sr *NDJSONStreamReader) Reset(r io.Reader) {
	if r == nil {
		sr.buf.Reset(nil)
		sr.reader.Reset(nil)
	} else {
		sr.buf.Reset(r)
		sr.reader.Reset(sr.buf)
	}
	sr.isEOF = false
	sr.latestLine = nil
	sr.latestLineReader.Reset(nil)
}
```
And then calling it from the `stream.Processor` like this:
```
func (p *Processor) HandleStream(ctx context.Context, ipRateLimiter *rate.Limiter, meta map[string]interface{}, reader io.Reader, report publish.Reporter) *Result {
	res := &Result{}

	sr, ok := p.streamReaderPool.Get().(*streamReader)
	if !ok {
		sr = &streamReader{}
		sr.NDJSONStreamReader = decoder.NewNDJSONStreamReader(reader, p.MaxEventSize)
	} else {
		sr.Reset(reader)
	}
	defer func() {
		sr.Reset(nil)
		p.streamReaderPool.Put(sr)
	}()
....
}
```

Reply:
Great idea, will do.